app/src/main/java/de/xikolo/controllers/settings/SettingsFragment.kt

```diff
@@ -41,9 +49,97 @@ class SettingsFragment : PreferenceFragment() {
+        if (key.equals(getString(R.string.preference_storage))) {
+            val newStoragePreference = sharedPreferences?.getString(getString(R.string.preference_storage), getString(R.string.settings_default_value_storage))!!
+            findPreference(getString(R.string.preference_storage)).summary = newStoragePreference
+
+            val newStorageType = StorageUtil.toStorageType(App.getInstance(), newStoragePreference)
+            var oldStorageType = StorageUtil.StorageType.INTERNAL
+            var oldStorage = StorageUtil.getInternalStorage(App.getInstance())
+            if (newStorageType == StorageUtil.StorageType.INTERNAL) {
+                oldStorageType = StorageUtil.StorageType.SDCARD
+                oldStorage = StorageUtil.getSdcardStorage(App.getInstance())!!
+            }
+
+            val fileCount = FileUtil.folderFileNumber(oldStorage)
+            if (fileCount > 0) {
+                val dialog = MigrationDialog.getInstance(activity, oldStorageType) {
+                    val progressDialog = ProgressDialog(activity)
+                    progressDialog.setTitle(R.string.dialog_storage_migration_title)
+                    progressDialog.setMessage(App.getInstance().getString(R.string.dialog_storage_migration_message))
+                    progressDialog.setCancelable(false)
+                    progressDialog.setCanceledOnTouchOutside(false)
+                    progressDialog.setProgressStyle(android.app.ProgressDialog.STYLE_HORIZONTAL)
+                    progressDialog.max = fileCount
+                    progressDialog.show()
+
+                    val migrationCallback = object : StorageUtil.StorageMigrationCallback {
+                        override fun onProgressChanged(count: Int) {
+                            activity.runOnUiThread { progressDialog.progress = count }
+                        }
+
+                        override fun onCompleted(success: Boolean) {
+                            activity.runOnUiThread {
+                                if (success)
+                                    ToastUtil.show(R.string.dialog_storage_migration_successful)
+                                else
+                                    ToastUtil.show(R.string.error_plain)
+                                progressDialog.dismiss()
+                            }
+                        }
+                    }
+
+                    if (newStorageType == StorageUtil.StorageType.INTERNAL) {
+                        StorageUtil.migrateAsync(StorageUtil.getSdcardStorage(App.getInstance())!!,
+                            StorageUtil.getInternalStorage(App.getInstance()),
+                            migrationCallback)
+                    } else {
+                        StorageUtil.migrateAsync(StorageUtil.getInternalStorage(App.getInstance()),
+                            StorageUtil.getSdcardStorage(App.getInstance())!!,
```

Comment:
`StorageUtil.getSdcardStorage()` can return `null` while `StorageUtil.getInternalStorage()` cannot. But `migrateAsync` requires two non-null storage locations, so `!!` asserts this. Actually the SDCard storage should always be non-null because migration can only be triggered from settings if there are more than one storage locations available.

Reply:
Nevertheless, it would be nice to make this code more robust and display a error message if this goes wrong. A null return for whatever reason would kill the whole app. Let's use Kotlin's Null Safety and try to avoid the usage of `!!` :)